File: api/writing_band.py
"""IELTS writing band helpers for evaluator portal."""

from __future__ import annotations

from typing import Any

CRITERIA_KEYS = ("task_achievement", "coherence", "lexical_resource", "grammar")


def round_half(value: float) -> float:
    return round(value * 2) / 2


def compute_overall_band(scores: dict[str, Any]) -> float:
    """Mean of four criteria, rounded to nearest 0.5 (IELTS-style)."""
    if any(key not in scores or scores[key] is None for key in CRITERIA_KEYS):
        raise ValueError("All four criteria scores are required.")
    values = [float(scores[key]) for key in CRITERIA_KEYS]
    return round_half(sum(values) / len(values))


def normalize_criteria_scores(raw: dict[str, Any] | None) -> dict[str, float] | None:
    if not raw:
        return None
    out: dict[str, float] = {}
    for key in CRITERIA_KEYS:
        if key not in raw or raw[key] is None:
            return None
        out[key] = float(raw[key])
    return out


def ai_scores_to_criteria(ai_scores: dict[str, Any] | None) -> dict[str, float] | None:
    """Map ai_scores / criteria_scores jsonb to rubric criteria."""
    if not ai_scores:
        return None
    nested = ai_scores.get("criteria_scores")
    if not isinstance(nested, dict):
        nested = ai_scores.get("criteria")
    if isinstance(nested, dict):
        ai_scores = nested
    mapped = {
        "task_achievement": ai_scores.get("task_achievement"),
        "coherence": ai_scores.get("coherence"),
        "lexical_resource": ai_scores.get("lexical_resource"),
        "grammar": ai_scores.get("grammar"),
    }
    if any(v is None for v in mapped.values()):
        return None
    return {k: float(v) for k, v in mapped.items()}

File: web/lib/writing-band.ts
/** IELTS writing criteria band math for evaluator portal. */

export type WritingHumanCriteriaScores = {
  task_achievement: number;
  coherence: number;
  lexical_resource: number;
  grammar: number;
};

export const WRITING_CRITERIA_KEYS: (keyof WritingHumanCriteriaScores)[] = [
  "task_achievement",
  "coherence",
  "lexical_resource",
  "grammar",
];

export const WRITING_CRITERIA_LABELS: Record<keyof WritingHumanCriteriaScores, string> = {
  task_achievement: "Task Achievement",
  coherence: "Coherence & Cohesion",
  lexical_resource: "Lexical Resource",
  grammar: "Grammatical Range & Accuracy",
};

export const WRITING_CRITERIA_DESCRIPTIONS: Record<
  keyof WritingHumanCriteriaScores,
  string
> = {
  task_achievement: "Addresses all parts of the task",
  coherence: "Logical organisation and linking",
  lexical_resource: "Range and precision of vocabulary",
  grammar: "Variety and accuracy of structures",
};

export function roundHalf(value: number): number {
  return Math.round(value * 2) / 2;
}

export function computeWritingOverallBand(
  scores: Partial<WritingHumanCriteriaScores>,
): number | null {
  const values = WRITING_CRITERIA_KEYS.map((key) => scores[key]);
  if (values.some((v) => v == null || Number.isNaN(v))) {
    return null;
  }
  const nums = values as number[];
  const mean = nums.reduce((sum, v) => sum + v, 0) / nums.length;
  return roundHalf(mean);
}

export function bandChipValues(min = 0, max = 9): number[] {
  const chips: number[] = [];
  for (let v = min; v <= max; v += 0.5) {
    chips.push(v);
  }
  return chips;
}

export function aiScoresToWritingCriteria(
  aiScores: Record<string, unknown> | null | undefined,
): WritingHumanCriteriaScores | null {
  if (!aiScores) return null;
  const nested =
    aiScores.criteria_scores && typeof aiScores.criteria_scores === "object"
      ? aiScores.criteria_scores
      : aiScores.criteria && typeof aiScores.criteria === "object"
        ? aiScores.criteria
        : null;
  const source =
    nested && typeof nested === "object"
      ? (nested as Record<string, unknown>)
      : aiScores;
  const task_achievement = source.task_achievement;
  const coherence = source.coherence;
  const lexical_resource = source.lexical_resource;
  const grammar = source.grammar;
  if (
    task_achievement == null ||
    coherence == null ||
    lexical_resource == null ||
    grammar == null
  ) {
    return null;
  }
  return {
    task_achievement: Number(task_achievement),
    coherence: Number(coherence),
    lexical_resource: Number(lexical_resource),
    grammar: Number(grammar),
  };
}

export function defaultWritingCriteriaFromReview(
  human: WritingHumanCriteriaScores | null | undefined,
  aiScores: Record<string, unknown> | null | undefined,
): WritingHumanCriteriaScores | null {
  if (human) return human;
  return aiScoresToWritingCriteria(aiScores);
}
